[KOGITO-7764] Timeout support

tiagodolphine opened the pull request:
<!-- Please don't forget your JIRA link -->
**JIRA:** https://issues.redhat.com/browse/KOGITO-7764

<!-- If you don't have a JIRA link, please provide a short description of what this PR does -->
**Description:**

<!-- Link to related PRs: -->

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributions doc](https://github.com/kiegroup/kogito-docs/blob/main/CONTRIBUTING.md)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [ ] The nav.adoc file has a link to this guide in the proper category
- [ ] The index.adoc file has a card to this guide in the proper category, with a meaningful description

<details>
<summary>
How to setup automatic cherry-pick to another branch ?
</summary>

The cherry-pick action allows to setup automatic cherry-pick from `main` to a specific branch.

To allow it, you will need to add the corresponding label with pattern: `backport-{RELEASE_BRANCH}`.  
For example, if a backport to branch `1.26.x` is needed, then the label should be `backport-1.26.x`.

Once the PR is merged, the action will retrieve the commit and cherry-pick it to the desired branch.

*NOTE: You can still add label after the merge and it should still be cherry-picked.*
</details>

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+:description: Using timeouts in {context}
+:keywords: kogito, workflow, serverless, timeout, timer, expiration
+
+When defining a state in a workflow you can set a timeout which is a property used to control the maximum time this state should take to be complete. In case of the time is elapsed and overdue, this state is considered timed-out, and the workflow moves to the next stated defined in the transition. The properties that you can use when defining a state timeout are described in the link:{spec_doc_ur ...
```
In any case we must use "timeouts" (with s) instead of timeout

tiagodolphine replied:
+1 incorporating you suggestion.

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+When defining a state in a workflow you can set a timeout which is a property used to control the maximum time this state should take to be complete. In case of the time is elapsed and overdue, this state is considered timed-out, and the workflow moves to the next stated defined in the transition. The properties that you can use when defining a state timeout are described in the link:{spec_doc_ur ...
+
+The event-based states can define a property `eventTimeout` which uses ISO 8601 data and time format to specify a duration of time, for instance, you can if you can represent 15 minutes like `PT15M`,  to represent 15 minutes or 2 days, 3 hours and 4 minutes `P2DT3H4M`.
```
I believe that here we can maybe include something like this:

"Event-based states can use the sub-property eventTimeout to configure the maximum time to wait for an event to arrive."

Since "timeouts" is a composed property. Another alternative could be to start saying in the initial paragrap that timeouts is a composed thing, but I think it's ok as it is, since we introduce the concept.

wmedvede replied:
thanks

wmedvede replied:
wait!

I believe this piece needs some more love:

"for instance, you can if you can represent 15 minutes like PT15M, to represent 15 minutes or 2 days, 3 hours and 4 minutes P2DT3H4M."

tiagodolphine replied:
+1, changing the text

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+[NOTE]
+====
+Event timeouts can not be defined as a specific point in time, but it should be an amount of time, a duration, which is considered to start when the referred state becomes active in the workflow.
```
I believe that if in the previous paragrpah we say that the event timeout must be a duration, we don't need this note.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+[IMPORTANT]
+====
+{product_name} currently, supports timeout only for *Callback* and *Switch* states with events, other states are going to support timeouts in the future releases.
```
Maybe...

"Current kogito version, has timeouts support only for the Callbakc and Swith states with events. Other states will be included in future releases."

wmedvede replied:
this change was not applied, not mandatory, but has the +1 from your side. Maybe you forgot.

wmedvede replied:
Ah,  see, was applied but you kept "Kogito currently, ....", ok, let's  @hmanwani-rh take a look. for me "Kogito currently, bla,bla " sounds strante.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+====
+
+=== Callback state timeout
+Callback state can be used when you need to publish an event to trigger an external service and waits an asynchronous response as an event, the callback. Once this event is consumed, the workflow continues the execution to the next state defined in the transition property. See more on xref:eventing/working-with-callbacks.adoc[Callback state in {context}].
```
Maybe

"Callback state can be used when you need to execute an action, for example invoke an external service, and wait for an asynchronous response in the form of an event, the callback"

wmedvede replied:
Maybe
"When the event arrives, it is consumed, and the workflow continues...."

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+=== Switch state timeout
+
+Switch states can be defined with link:{spec_doc_url}#switch-state-event-conditions[eventConditions] property, where the workflow execution waits to make a decision depending on the events to be consumed and matched, defined through link:{spec_doc_url}#event-definition[event definition].
```
Maybe here we can say something like 
"The switch state can be used when you need too..." 
to get it more similar to the previous paragraph, and then the rest

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+Switch states can be defined with link:{spec_doc_url}#switch-state-event-conditions[eventConditions] property, where the workflow execution waits to make a decision depending on the events to be consumed and matched, defined through link:{spec_doc_url}#event-definition[event definition].
+
+In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property in the switch state, as showed in the <<switch-state, example>>.
```
in the `defaultCondition` of the switch state, as you can see in the following example.

Maybe the link "<<switch-state, example>>." is in needed, since it's in the line below

wmedvede replied:
not sure if this was applied, I tried to say that the link is not needed since the example comes in the next lines. But np.

tiagodolphine replied:
+1, anyway I think it is great to have the link the text on the link remains the same.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+n order to deploy a work that contains timeouts or any other timer-based action, it is necessary to have Job Service running in your environment, which is an external service responsible for workflows timers, see the <<job-service, section>> for more information.
+In the <<timeout-example, timeout example>> you can see in details how to set up a knative infrastructure with the workflow and job service running.
+
+[#job-service]
```
Jobs Service or Job Service?, both are good, but I think we don't have a fixed name yet. Sometimes we say job, other times we say jobs.
In the kogito business automation documentation we say "Jobs Service"

ricardozanini replied:
cc @hmanwani-rh

ricardozanini replied:
cc @kaldesai now :P

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+Suppose the workflow service is not configured to use job service or there is no such service running. In that case, all timer-related actions use an embedded in-memory implementation of job service, which should not be used in production, since when the application shutdown, all timers are lost, which in a serverless architecture is a very common behavior with the scale to zero approach. That sa ...
+
+The main goal of the Job Service is to work with only active jobs. The Job Service tracks only the jobs that are scheduled and that need to be executed. When a job reaches a final state, the job is removed from the Job Service. All job information and transition states are sent to the Kogito Data Index Service where they can be indexed and made available for GraphQL queries.
```
When configured, all the jobs information and status changes are sent to the Kogito Data Index Service, where they can be indexed and made available for GraphQL queries.

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+==== PostgreSQL
+
+Using PostgreSQL is the recommended database to be used with job service where it is already configured with Flyway for the schema creation, in this way all tables are created by the service, in case you need to externally control the database schema, you can check the SQL scripts in (link) and apply them.
```
May be we can rephrase the start of this paragraph to something like this.

PostgreSQL is the recommended database to use with job service.
Additionally, the jobs service provides an initialization procedure that integrates Flyway for the database initialization, in this way, all the tables, etc., are automatically created by the service.

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+==== PostgreSQL
+
+Using PostgreSQL is the recommended database to be used with job service where it is already configured with Flyway for the schema creation, in this way all tables are created by the service, in case you need to externally control the database schema, you can check the SQL scripts in (link) and apply them.
+You need to set the proper configuration parameters when starting job service. The example covers running PostgreSQL as a Kubernetes deployment but you can run it the way it fits in your environment, the important part is to set all the configuration parameters points to your running instance of PostgreSQL.
```
Maybe

"The example shows how to run PostgreSQL as a Kubernetes deployment but you can run it the way it fits in your ....."

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+==== Ephemeral
+Alternatively, there is an in-memory database implementation that does not require any external database configuration, similar to the workflow runtime embedded in-memory job service (link), it can be used for testing and development purposes, but it is not recommended for production, since all jobs are lost in case of a service restart or failure.
+
+=== Job Service communication
```
Don't understood too much this paragraph, but yeah, maybe it's not ready.

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+[#timeout-example]
+== Timeout showcase example
+
+In the link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[serverless-workflow-timeouts-showcase] you can see a complete architecture, containing a serverless workflow application with timeouts configured alongside Job service running on Knative.
```
Maybe start with something like this:

In the serverless-workflow-timeouts-showcase you can see an end-to-end example that contains....... 
integrates.....

etc

wmedvede replied:
sometimes we write job service, Job Service, Job service. Wondering if we should try to standarize always uppercased "Job Serivce" maybe?

tiagodolphine replied:
I'm fine, let's keep Job Service and let @hmanwani-rh give her suggestion later on the review.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+In the link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[serverless-workflow-timeouts-showcase] you can see a complete architecture, containing a serverless workflow application with timeouts configured alongside Job service running on Knative.
+
+There are two workflows to showcase timeouts usage: `Callback` and `Switch` states.
```
There are two workflows that showcase the timeouts usage in the: Callback and Switch states.

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+=== Callback workflow
+
+It is a simple workflow where once the execution reaches the callback state it waits for the event `callbackEvent`to be received and continue the execution.
```
It is a simple workflow where, (a comma here I believe)

wmedvede replied:
"event `callbackEvent`to arrive" instead of "be received"

wmedvede replied:
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+}
+----
+
+A timeout is configured with a maximum time 30 seconds to be waited by the workflow to receive the callbackEvent, in case it does not arrive in time, the execution moves, and the eventData variable remains null.
```
The timeout is configured... 

Maybe to show the callback sate json element instead of the event can help.?

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+=== Switch workflow
+
+The switch example is similar to the callback but once the execution reaches the state, it waits for once of the two configured events, `visaDeniedEvent` or `visaApprovedEvent`, to be consumed, see the <<switch-state, switch state definition>>.
```
it waits for ONE of the two configured events, visaDeniedEvent or visaApprovedEvent, to ARRIVE, see the [switch state definition](https://deploy-preview-198--tangerine-crostata-04a6bf.netlify.app/serverlessworkflow/latest/core/timeouts-support.html#switch-state).

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+Whenever one of the events is consumed, before the configured timeout, the workflow execution moves to the next state defined in the `transition`.
+
+If none of the events arrive before 30 seconds of timeout, the workflow then moves to the state defined in `defaultCondition' transition.
```
Maybe:
If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition' transition.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+=== Running the example
+
+In order to run the example you need to have a kubernetes or Openshift cluster running with Knative configured. In the example minikube is being used you can follow the steps described in the example's link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[readme].
```
To run the example you must have access too a kubernetes....

wmedvede replied:
For simplicity, the example use minikube... bla, bla

tiagodolphine replied:
+1

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+In order to run the example you need to have a kubernetes or Openshift cluster running with Knative configured. In the example minikube is being used you can follow the steps described in the example's link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[readme].
+
+All the descriptor files used to deploy the example infrastructure are present in the kubernetes folder, and for the workflow application, the descriptors are placed on the target/kubernetes, generated after the build.
```
Maybe here we can add a [NOTE] ?
And spit a bit phrase, like "The descriptor related to the showcase-databse and jobs service deployment are located at....
The descriptors related to the application....

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+All the descriptor files used to deploy the example infrastructure are present in the kubernetes folder, and for the workflow application, the descriptors are placed on the target/kubernetes, generated after the build.
+
+The diagram below describes how is the architecture of the example to be deployed on kubernetes and Knative.
```
The following diagram shows the example's architecture when it is deployed in the Kubernetes + Knative infrastructure.

tiagodolphine replied:
+1

tiagodolphine replied:
+1

wmedvede reviewed (commented):
Hi @tiagodolphine, I added some comments since I was invited to review. I understand it's a draft, but some might still be useful.

tiagodolphine commented:
> Hi @tiagodolphine, I added some comments since I was invited to review. I understand it's a draft, but some might still be useful.

Thanks for the review @wmedvede I'll take a look at your suggestions and apply them, I think we are ready for review, so I moved this from draft to ready.

ricardozanini commented:
@tiagodolphine @wmedvede we don't need to backport since it's a community only feature, correct?

tiagodolphine commented:
> @tiagodolphine @wmedvede we don't need to backport since it's a community only feature, correct?

I'm not sure, I understood that there will be a doc freeze for DP2, if we don't backport, in DP guides there will be no reference to timeout, even though it is community only. Anyway up to us, tbh I don't know the best approach to mention community only stuff on product docs. wdyt @ricardozanini @hmanwani-rh ?

ricardozanini commented:
@tiagodolphine the backport is to bring this to product. 1.27 is our release train for OpenShift Serverless. That's why I think we should not backport. It will be in the next version, so we should be fine. Maybe backport to 1.28. @radtriste

radtriste commented:
1.28.x is going to be created soon. If you want to backport to it, then just add the `backport-1.28.x` label

ricardozanini commented:
> 1.28.x is going to be created soon. If you want to backport to it, then just add the `backport-1.28.x` label

In that case, we can merge this before branching if possible.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+   "functionRef": {
+     "refName": "callbackFunction",
+     "arguments": {
+       "input": "${\"callback-state-timeouts: \" + $WORKFLOW.instanceId + \" has executed the callbackFunction.\"}"
```
I think this argument can be an object instead. This way, you avoid adding these scape chars and educate the users.

tiagodolphine replied:
yeah, @wmedvede can you update the example in this way we align with the guide

wmedvede replied:
Hi @ricardozanini, I missed this comment, but what do you mean with an object?

Let me explain:
The idea here is that the invoked function is sysout, why? to make things simple, etc., but also during the review process it was also suggested to remove a kind of "more real service" invocation I had, that in the end did the system out.
So, the idea is, that this two SWs are intended to show specifically how to use the timeouts and to provide a quickly way to see understand the code, execute, and see the results in the system output. Let's say a one-to-one matching.

Not sure if I've got your point, but to have an object as argument, I need to create it first, so, or we pass it to the SW as parameter, or we use an inject state, or we create it somehow in another point of the SW, etc.
But with this options, I loose the end-to-end mapping from the SW definition to the system output, and the simplicity of reading the flow.

What do you suggest?

ricardozanini replied:
Oh ok, this is used in sysout, I missed that. This is fine.

ricardozanini reviewed (commented):
I will do a second round review after @wmedvede's comments are incorporated. 
Can I ask you to please look at the [CNCF Doc](https://github.com/kiegroup/kogito-docs/blob/main/serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc) and update there too?

tiagodolphine commented:
@ricardozanini I already applied Walter's comments.

Thanks.

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+=== Callback state timeout
+Callback state can be used when you need to publish an event to trigger an external service and waits an asynchronous response as an event, the callback. Once this event is consumed, the workflow continues the execution to the next state defined in the transition property. See more on xref:eventing/working-with-callbacks.adoc[Callback state in {context}].
+
+Since the callback event halts the execution util the event is consumed, you can define an `eventTimeout` for it, and in case the event does not arrive in the defined duration time, the workflow continues the execution moving to the next state defined in the transition, see the <<callback-state, example>>.
```
thanks

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+=== Deploying a timed-based workflow
+
+n order to deploy a work that contains timeouts or any other timer-based action, it is necessary to have Job Service running in your environment, which is an external service responsible for workflows timers, see the <<job-service, section>> for more information.
```
The paragraph should starts with "i" in lowercase.

I think we should say "you can see the details" insead of "in deatils", maybe @hmanwani-rh can help with the most correct form.

tiagodolphine replied:
+1

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,359 @@
+
+==== Knative Eventing
+
+You can configure job service to consume and publish events through knative setting some configuration parameters in your environment.
```
Maybe we can start with something like this:

"To rely the communication between the job service and the workflows runtime to the knative eventing system,  you must provide a set of configurations."

"The Job Service configuration is done through the deployment descriptor shown in the example."

wmedvede replied:
"is done through an addon, (<-comma) which is responsible for publishing and consuming events related to timers.

wmedvede replied:
When you run the workflow as a knative service, you must add "bla bla bla" to your project and provide the proper configuration.

tiagodolphine replied:
+1

wmedvede commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,386 @@
+timeouts-showcase         http://timeouts-showcase.default.10.105.86.217.sslip.io         timeouts-showcase-00001         3m50s   3 OK / 3     True
+----
+
+=== Using the show UI
```
"Using the showcase" instead of "show"

tiagodolphine replied:
+1

wmedvede commented:
Hi @tiagodolphine , thanks for the modifications. I've added a couple of small comment. I'll let this in stand-by from my side until we get feedback/suggestions from @hmanwani-rh  and the rest of the team.

tiagodolphine commented:
Thanks for the review and suggestions @wmedvede, I already updated the PR.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+All the properties you can use to configure state timeouts are described in the link:{spec_doc_url}#event-timeout-definition[Serverless Workflow specification].
+
+Event-based states can use the sub-property `eventTimeout` to configure the maximum time to wait for an event to arrive.
+This property uses `ISO 8601` data and time standard to specify a duration of time. It follows the format `PnDTnHnMn.nS` with days considered to be exactly 24 hours. For instance,
```
```suggestion
This property uses link:https://en.wikipedia.org/wiki/ISO_8601[`ISO 8601` data and time standard] to specify a duration of time. It follows the format `PnDTnHnMn.nS` with days considered to be exactly 24 hours. For instance,
```

You may consider adding this link as a page attribute.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+
+[IMPORTANT]
+====
+{product_name} currently, has timeouts support only for *Callback* and *Switch* states with events. Other states will be included in the future releases.
```
Please add this information to the CNCF guide.

tiagodolphine replied:
+1

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+Suppose the workflow service is not configured to use job service or there is no such service running. In that case, all timer-related actions use an embedded in-memory implementation of job service, which should not be used in production, since when the application shutdown, all timers are lost, which in a serverless architecture is a very common behavior with the scale to zero approach. That sa ...
+
+The main goal of the Job Service is to work with only active jobs. The Job Service tracks only the jobs that are scheduled and that need to be executed. When a job reaches a final state, the job is removed from the Job Service.
+When configured, all the jobs information and status changes are sent to the Kogito Data Index Service, where they can be indexed and made available by GraphQL queries.
```
@cristianonicolai do we have a plan to write a Data Index guide? I think it should be prioritized since we mention it here without a link. It might confuse users. :/

tiagodolphine replied:
I'll insert a sentence to mention it will be included in the future releases

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+
+==== PostgreSQL
+
+PostgreSQL is the recommended database to use with job service**.**
```
What's the role of this `**.**`?

tiagodolphine replied:
lol, it was added by mistake, removing it

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+PostgreSQL is the recommended database to use with job service**.**
+Additionally, it provides an initialization procedure that integrates link:https://flywaydb.org[Flyway] for the database initialization. It automatically controls the database schema, in this way all tables are created by the service.
+
+In case you need to externally control the database schema, you can check the Flyway SQL scripts in link:{kogito_apps_url}/jobs-service/jobs-service-postgresql/src/main/resources/db/migration[migration] and apply them.
```
Is this `kogito_apps_url` versioned?

tiagodolphine replied:
it is pointing to main branch,  `kogito_apps_url: https://github.com/kiegroup/kogito-apps/tree/main`

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+
+[NOTE]
+====
+The Job Service does not execute a job but triggers a callback that might be an HTTP request or a Cloud Event that is managed by the configured jobs addon see(link addons config) in the workflow application.
```
I think we are missing the link :D

tiagodolphine replied:
yeahhh, thanks

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,388 @@
+mp.messaging.outgoing.kogito-job-service-job-request-events.connector=quarkus-http
+mp.messaging.outgoing.kogito-job-service-job-request-events.url=${K_SINK:http://localhost:8280/jobs/events}
+mp.messaging.outgoing.kogito-job-service-job-request-events.method=POST
+----
```
I think it's worth explaining that the `K_SINK` variable is injected by Knative Eventing and add alink here to our Knative Eventing integration guide.

tiagodolphine replied:
+1

ricardozanini reviewed (approved):
LGTM

tiagodolphine commented on serverlessworkflow/antora.yml:
```diff
@@ -36,8 +37,9 @@ asciidoc:
     kogito_devservices_imagename: registry.redhat.io/openshift-serverless-1-tech-preview/logic-data-index-ephemeral-rhel8
     kogito_examples_repository_url: https://github.com/kiegroup/kogito-examples
-    kogito_sw_examples_url: https://github.com/kiegroup/kogito-examples/tree/main/serverless-workflow-examples
+    kogito_sw_examples_url: https://github.com/kiegroup/kogito-examples/tree/{kogito_upstream_branch}/serverless-workflow-examples
```
@ricardozanini do you know how to reference an attribute inside another attribute? from the docs it seems it should be in this way, but it didnt work 🤔 
in the and I can hardcode the version in the URL if there is no solution...

ricardozanini replied:
Unfortunately, Antora doesn't support it. The CI will change that once we do a release. cc @radtriste 

Just hardcode the version for now.

radtriste replied:
yep, the kgoito prepare release pipeline will change the kogito exampels URL

tiagodolphine replied:
> I think we can leave this as main, it will be changed to 1.27.x when forking, right?

yeah, good point @MarianMacik , the timeout example and job-service will be available just in community so it wont be cherry picked to 1.27.x

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+
+The switch state can be used when you need to take an action based on conditions, defined with the link:{spec_doc_url}#switch-state-event-conditions[eventConditions] property, where the workflow execution waits to make a decision depending on the events to be consumed and matched, defined through link:{spec_doc_url}#event-definition[event definition].
+
+In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property of the switch state, as tou can see in the <<switch-state, example>>.
```
```suggestion
In this situation, you can define an event timeout, that controls the maximum time to wait for an event to match the conditions, if this time is expired, the workflow moves to the state defined in the `defaultCondition` property of the switch state, as you can see in the <<switch-state, example>>.
```

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+
+=== Callback workflow
+
+It is a simple workflow, where once the execution reaches the callback state it waits for the event `callbackEvent`to arrive and continue the execution.
```
```suggestion
It is a simple workflow, where once the execution reaches the callback state it waits for the event `callbackEvent` to arrive and continue the execution.
```

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+
+If any of the configured events arrives before the timeout is overdue, the workflow execution moves to the next state defined in the corresponding `transition`.
+
+If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition' transition.
```
```suggestion
If none of the events arrive before the timeout is overdue, the workflow then moves to the state defined in `defaultCondition` transition.
```

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+For simplicity, the example uses minikube, you can follow the steps described in the example's link:{kogito_sw_examples_url}/serverless-workflow-timeouts-showcase[readme].
+
+[NOTE]
+----
+All the descriptor files used to deploy the example infrastructure are present in the example.
+----
```
For NOTE it needs to be
```suggestion
====
All the descriptor files used to deploy the example infrastructure are present in the example.
====
```
I believe.

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+.Timeout workflow showcase UI
+image::core/timeout-switch-wokflow-ui.png[]
+
+You can create new workflow instances and interact with them to complete, or simply wait the timeout to be triggered to check it working.
```
```suggestion
You can create new workflow instances and interact with them to complete, or simply wait for the timeout to be triggered to check it's working.
```

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/core/timeouts-support.adoc:
```diff
@@ -0,0 +1,407 @@
+[]
+----
+
+As you can see there are no active workflow instances, indicating the timout was executed and the created instance was completed.
```
```suggestion
As you can see there are no active workflow instances, indicating the timeout was executed and the created instance was completed.
```

MarianMacik reviewed (approved):
Looks good, just minor cosmetic issues. For URL, we are on main here, so I think we should leave it pointing to main and when the fork is done, only then the URLs should be updated.

kaldesai commented on serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:
```diff
@@ -229,7 +231,8 @@ Alternatively, you can use xref:core/understanding-workflow-error-handling.adoc[
 == Timeouts
 
-{product_name} does not support Timeouts feature, however, it will be implemented in a future release.
+{product_name} has a limited support of Timeouts feature, that covers only *Callback* and *Switch* states with events.
```
```suggestion
{product_name} has a limited support of Timeouts feature, that covers only *Callback* and *Switch* states with events.
```
```suggestion
{product_name} has limited support for the Timeouts feature, which covers only *Callback* and *Switch* states with events.
```

kaldesai reviewed (approved):
@tiagodolphine Very minor suggestions. Otherwise, your content looks good to me. Thank you!
